yang-module-versioning/draft-ietf-netmod-yang-module-versioning.xml

```diff
@@ -216,7 +216,7 @@
     NBC changes.</t>
     <t>If a revision of a YANG module contains changes, relative to the preceding revision in the revision history,
     that do not conform to the module update rules defined in <xref target="bc_update_rules"/>, then a
-    "rev:non-backwards-compatible-revision" extension statement MUST be added as a substatement to the "revision" statement.</t>
+    "rev:non-backwards-compatible-revision" extension statement SHOULD be added as a substatement to the "revision" statement. If the YANG module changes are non-backwards-compatible but there is no change in the values accepted by the servers implementing the YANG module,  then a "rev:non-backwards-compatible-revision" extension statement MUST NOT be added as a substatement to the "revision" statement.</t>
```

Comment:
Although maybe that first change of MUST to SHOULD, should instead be rework as "MUST, unless" (or MUST, with the followign exception" or something like that.

Reply:
> We can't say "MUST NOT" in the last part (since we always allow NBC to be marked). Perhaps this:

What about SHOULD NOT.

Reply:
> Although maybe that first change of MUST to SHOULD, should instead be rework as "MUST, unless" (or MUST, with the followign exception" or something like that.

MUST unless is a SHOULD afaik.

Reply:
The wording here is tricky. We want to lean towards strongly requiring the NBC change marker when the author considers the change as NBC, but if the author considers it BC, then they don't *need* the nbc marker, but they are *allowed* to put it (strictly speaking).

I'd lean towards making the first part a "MUST, except when" to keep it strong and bound when it can be left off. The last one could be "SHOULD NOT" though.

Reply:
> I'd lean towards making the first part a "MUST, except when" to keep it strong and bound when it can be left off. 

So you're worried that if we use SHOULD, instead of "MUST, except when ...", that it'll be too soft? I'm just not used to seeing that in RFCs, usually it's SHOULD with an explanation of why it's not MUST, but it's possible I haven't been paying attention.  Let's discuss in the weekly meeting, should be quick.

Reply:
A) then a "rev:non-backwards-compatible-revision" extension statement SHOULD be added as a substatement to the "revision" statement. If the YANG module changes are non-backwards-compatible but there is no change in the values accepted by the servers implementing the YANG module,  then a "rev:non-backwards-compatible-revision" extension statement MUST NOT be added as a substatement to the "revision" statement.</t>

B)  If a revision of a YANG module contains changes, relative to the preceding revision in the revision history,
    that do not conform to the module update rules defined in <xref target="bc_update_rules"/>, then a "rev:non-backwards-compatible-revision" extension statement MUST be added as a substatement to the "revision" statement, unless there is no in the values accepted by the servers implementing the YANG module, in which case a "rev:non-backwards-compatible-revision" extension statement is OPTIONAL.

C)  If a revision of a YANG module contains changes, relative to the preceding revision in the revision history,
    that do not conform to the module update rules defined in <xref target="bc_update_rules"/>, then a "rev:non-backwards-compatible-revision" extension statement MUST be added as a substatement to the "revision" statement, unless all previously accepted values are still accepted by the servers implementing the YANG module, in which case a "rev:non-backwards-compatible-revision" extension statement is OPTIONAL.

D)  If a revision of a YANG module contains changes, relative to the preceding revision in the revision history,
    that do not conform to the module update rules defined in <xref target="bc_update_rules"/>, then a "rev:non-backwards-compatible-revision" extension statement SHOULD be added as a substatement to the "revision" statement. The exception is when all previously accepted values are still accepted by the servers implementing the YANG module, in which case a "rev:non-backwards-compatible-revision" extension statement is OPTIONAL.

Proposal on the call -> go with wording C

Reply:
Per import changes pushed by @rgwilton, afaik  "rev:non-backwards-compatible-revision"  ->  "rev:non-backwards-compatible"

Reply:
Went ahead with wording C, another pair of eyes please.

Reply:
We could consider changing "unless all previously accepted values are still accepted" to "unless all previously accepted combinations of configuration statements are still accepted".  It isn't actually enough to just accept all the same values for an individual leaf. A server needs to accept all complete configurations that it used to accept. For example - if two pre-existing leafs are moved into 2 cases of a new choice, then they are newly mutually exclusive. On their own, each leaf still accepts all the same old values as before. But the combinations of the two leafs has changed.

Reply:
There is also a corner case here in our current wording: it does allow a standards model to have a choice or case renamed. That doesn't affect the values accepted, but does affect any module trying to augment the standard model. This case isn't so applicable to vendor modules.
Similarly our wording allows a grouping to be renamed.
Not sure it is worth trying to contain these cases in our wording or not, but we *could* consider adding something about not affecting augmentations or includes by other modules?

Reply:
> For example - if two pre-existing leafs are moved into 2 cases of a new choice, then they are newly mutually exclusive. On their own, each leaf still accepts all the same old values as before. But the combinations of the two leafs has changed.

I'm not sure I get this. We've gone from 2 leafs to 2 cases of a new choice, so to 1 leaf then? So you're saying each of the previously supported values is still accepted so that's ambiguous?

> There is also a corner case here in our current wording: it does allow a standards model to have a choice or case renamed. That doesn't affect the values accepted, but does affect any module trying to augment the standard model. This case isn't so applicable to vendor modules.

The current wording is geared towards the IP address pattern change example. Not sure how best to put all these cases in this paragraph.  Maybe adding the IP address example would go some ways in making it clearer what the intent is.

Reply:
For the first example. 

Before:
leaf a { type string; }
leaf b { type int; }

After:
choice foo {
    case first {
        leaf a { type string; }
    }
    case second {
        leaf b { type int; }
    }
}

Leaf a still allows all the same values as before.
Leaf b still allows all the same values as before.

You could argue that "all previously accepted values are still accepted by the servers implementing the YANG module" is satisfied (if you look at one element at a time).

But the combination of:
a = hello
b = 5
is no longer accepted.

Reply:
I'm not too sure we need to cover these cases I'm bringing up though. *If* we want to cover them, then here is a possible C':
If a revision of a YANG module contains changes, relative to the preceding revision in the revision history,
that do not conform to the module update rules defined in , then a "rev:non-backwards-compatible-revision" extension statement MUST be added as a substatement to the "revision" statement, unless:
    1) all previously accepted combinations of configuration statements are still accepted by the servers implementing the YANG module, and
    2) the changes to the module do not impact augmentations or imports of the module by other modules,
in which case a "rev:non-backwards-compatible-revision" extension statement is OPTIONAL.

Reply:
I'm good with something along those lines. Also think having an example would help.

Reply:
C'+:
If a revision of a YANG module contains changes, relative to the preceding revision in the revision history,
that do not conform to the module update rules defined in , then a "rev:non-backwards-compatible" extension statement MUST be added as a substatement to the "revision" statement, unless:
1) all previously accepted combinations of configuration statements are still accepted by the servers implementing the YANG module, and
2) the changes to the module do not impact augmentations or imports of the module by other modules (e.g. choice statements and groupings are not renamed),
in which case a "rev:non-backwards-compatible" extension statement is OPTIONAL. For example, if a vendor YANG model allowed a range for an uint8 leaf of 0..50, but the server only ever allowed values 1..49 to be committed, then a correction of the range in the YANG model to 1..49 would not require the "rev:non-backwards-compatible" extension.